envoy/stats/stats_macros.h

```diff
@@ -152,16 +152,41 @@ static inline std::string statPrefixJoin(absl::string_view prefix, absl::string_
  */
 #define MAKE_STATS_STRUCT(StatsStruct, StatNamesStruct, ALL_STATS)                                 \
   struct StatsStruct {                                                                             \
+    using StatNameType = StatNamesStruct;                                                          \
     StatsStruct(const StatNamesStruct& stat_names, Envoy::Stats::Scope& scope,                     \
                 Envoy::Stats::StatName prefix = Envoy::Stats::StatName())                          \
         : stat_names_(stat_names)                                                                  \
               ALL_STATS(MAKE_STATS_STRUCT_COUNTER_HELPER_, MAKE_STATS_STRUCT_GAUGE_HELPER_,        \
                         MAKE_STATS_STRUCT_HISTOGRAM_HELPER_,                                       \
                         MAKE_STATS_STRUCT_TEXT_READOUT_HELPER_,                                    \
                         MAKE_STATS_STRUCT_STATNAME_HELPER_) {}                                     \
-    const StatNamesStruct& stat_names_;                                                            \
+    const StatNameType& stat_names_;                                                               \
     ALL_STATS(GENERATE_COUNTER_STRUCT, GENERATE_GAUGE_STRUCT, GENERATE_HISTOGRAM_STRUCT,           \
               GENERATE_TEXT_READOUT_STRUCT, GENERATE_STATNAME_STRUCT)                              \
   }
 
+/**
+ * LazyInit$StatsStruct is a wrapper with creator of the actual "$StatsStruct"
+ * structure.
+ * It instantiate a $StatsStruct struct when any data member is referenced.
+ * See https://github.com/envoyproxy/envoy/issues/23575 for more details.
+ */
+template <typename StatsStructType> struct LazyInitStats {
```

Comment:
not really a macro, though I can see why it's convenient to put it here. Maybe for now just put this in cluster.h as its only use?

Reply:
it's not a macro -- can you just put it in its own file? You can leave it in envoy/stats/... and depend on it from upstream/upstram.h so it won't be a difficult change.

Reply:
Also this should be namespaced. Macros are not namespaced. Then it will be Stats::LazyInit rather than LazyInitStats, which I think would be more consistent if you want to put it in stats infrastructure.